### PR TITLE
Simplify optimization Logic

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR attempts uniformize between automatic and manual optimization.   

It allows resolve a bug where optimizer_step from model wasn't called with LightningOptimizer.

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1236,15 +1237,11 @@ def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx,
...
-        if on_tpu and TPU_AVAILABLE:
-            xm.optimizer_step(optimizer, optimizer_args={'closure': optimizer_closure, **kwargs})
-
-        elif self.trainer.amp_backend is not None:
-            self.trainer.precision_connector.backend.optimizer_step(
-                self.trainer, optimizer, optimizer_closure)
-
-        else:
-            optimizer.step(closure=optimizer_closure, *args, **kwargs)
+        if not isinstance(optimizer, LightningOptimizer):
```
Comment (SeanNaren) [inline pytorch_lightning/core/lightning.py:1240]: Neat, what about a static function within LightningOptimizer to do this? Just throwing out ideas!

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/core/optimizer.py [hunk 2]
```diff
@@ -60,17 +60,34 @@ def __init__(self,
...
+    def accumulate_grad_batches(self):
+        return self._accumulate_grad_batches
+
+    @accumulate_grad_batches.setter
+    def accumulate_grad_batches(self, accumulate_grad_batches):
+        if self._automatic_optimization is None or self._automatic_optimization:
+            raise MisconfigurationException(
+                'In automatic optimization, `make_optimizer_step` should be None. '
+                'This option is enabled only with Trainer(automatic_optimization=False) '
+                'Currently, all optimizers follow the Trainer(accumulate_grad_batches=x) logic')
```
Comment (Borda) [inline pytorch_lightning/core/optimizer.py:76]:
```suggestion
                'In automatic optimization, `make_optimizer_step` should be None.'
                ' This option is enabled only with `Trainer(automatic_optimization=False)`'
                ' Currently, all optimizers follow the `Trainer(accumulate_grad_batches=x)` logic')
```

File: pytorch_lightning/core/optimizer.py [hunk 3]
```diff
@@ -60,17 +60,34 @@ def __init__(self,
...
+                self._optimizer_idx = opt_idx
+                break
 
     def _accumulated_batches_reached(self):
-        if self._use_accumulate_grad_batches_from_trainer:
-            accumulate_grad_batches = self._trainer.accumulate_grad_batches
-        else:
-            accumulate_grad_batches = self._accumulate_grad_batches
-        return (self._trainer.batch_idx + 1) % accumulate_grad_batches == 0
+        if self._accumulate_grad_batches is None:
```
Comment (Borda) [inline pytorch_lightning/core/optimizer.py:88]:
```suggestion
        if self.accumulate_grad_batches is None:
```

File: pytorch_lightning/core/optimizer.py [hunk 4]
```diff
@@ -60,17 +60,34 @@ def __init__(self,
...
+            return self._trainer.train_loop._accumulated_batches_reached()
+        return (self._trainer.batch_idx + 1) % self._accumulate_grad_batches == 0
```
Comment (Borda) [inline pytorch_lightning/core/optimizer.py:90]:
```suggestion
        return (self._trainer.batch_idx + 1) % self.accumulate_grad_batches == 0
```

File: pytorch_lightning/core/optimizer.py [hunk 5]
```diff
@@ -184,30 +227,19 @@ def dis_closure():
...
             make_optimizer_step = not self._should_accumulate
-
-        trainer = self._trainer
-        optimizer = self._optimizer
+        else:
+            if self._automatic_optimization:
+                raise MisconfigurationException(
+                    'In automatic optimization, `make_optimizer_step` should be None. '
+                    'This option is enabled only with Trainer(automatic_optimization=False) '
+                    'Currently, all optimizers follow the Trainer(accumulate_grad_batches=x) logic')
```
Comment (Borda) [inline pytorch_lightning/core/optimizer.py:235]:
```suggestion
                    'In automatic optimization, `make_optimizer_step` should be None.'
                    ' This option is enabled only with `Trainer(automatic_optimization=False)`'
                    ' Currently, all optimizers follow the `Trainer(accumulate_grad_batches=x)` logic')
```
lets add test for all there exceptions...

File: pytorch_lightning/core/optimizer.py [hunk 6]
```diff
@@ -79,6 +96,32 @@ def _should_accumulate(self):
...
+        optimizer = self._optimizer
+        model = trainer.get_model()
+
+        if trainer.on_tpu:
+            with trainer.profiler.profile(profiler_name):
+                xm.optimizer_step(optimizer, optimizer_args={'closure': closure, **kwargs})
+
+        elif trainer.amp_backend is not None:
+            trainer.precision_connector.backend.optimizer_step(
+                trainer, optimizer, closure)
```
Comment (Borda) [inline pytorch_lightning/core/optimizer.py:110]:
```suggestion
            trainer.precision_connector.backend.optimizer_step(trainer, optimizer, closure)
```

File: pytorch_lightning/core/optimizer.py [hunk 7]
```diff
@@ -184,30 +227,19 @@ def dis_closure():
...
-
-            # perform zero grad
-            optimizer.zero_grad()
+            self.__optimizer_step(*args, closure=closure, profiler_name=profiler_name, **kwargs)
         else:
             # make sure to call optimizer_closure when accumulating
-            with trainer.profiler.profile("closure"):
-                with trainer.train_loop.block_ddp_sync_behaviour():
-                    closure()
+            with self._trainer.profiler.profile("closure"), self._trainer.train_loop.block_ddp_sync_behaviour():
```
Comment (Borda) [inline pytorch_lightning/core/optimizer.py:241]: shall sec which closure - closure for what optimizer?

File: pytorch_lightning/plugins/native_amp.py [hunk 8]
```diff
@@ -69,6 +69,11 @@ def optimizer_step(self, trainer, optimizer, closure):
         # TODO: pass the closure to the step ASAP
         with trainer.profiler.profile("closure"):
             closure()
+
+        if not self.trainer.train_loop.automatic_optimization:
+            trainer.scaler.unscale_(optimizer)
+            trainer.call_hook("on_after_backward")
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/plugins/native_amp.py:75]: does it mean that this hook is called only in manual optim?

File: pytorch_lightning/trainer/configuration_validator.py [hunk 9]
```diff
@@ -69,6 +69,25 @@ def __verify_train_loop_configuration(self, model):
...
 
+        has_optimizer_step = is_overridden('optimizer_step', model)
+        enable_pl_optimizer = self.trainer._enable_pl_optimizer
+        automatic_optimization = self.trainer.train_loop.automatic_optimization
+        if has_optimizer_step and not enable_pl_optimizer and automatic_optimization:
+            log.warn(
+                "When overriding `LightningModule` optimizer_step with "
+                "Trainer(..., enable_pl_optimizer=False, automatic_optimization=True, ...), "
+                "we won t be calling `.zero_grad` we can't assume when you call your optimizer.step(). "
+                "For Lightning to take care of it, please use Trainer(enable_pl_optimizer=True)."
```
Comment (Borda) [inline pytorch_lightning/trainer/configuration_validator.py:80]:
```suggestion
                "When overriding `LightningModule` optimizer_step with"
                " `Trainer(..., enable_pl_optimizer=False, automatic_optimization=True, ...)`,"
                " we won t be calling `.zero_grad` we can't assume when you call your `optimizer.step()`."
                " For Lightning to take care of it, please use `Trainer(enable_pl_optimizer=True)`."
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 10]
```diff
@@ -69,6 +69,25 @@ def __verify_train_loop_configuration(self, model):
...
+            )
+
+        going_to_accumulate_grad_batches = self.trainer.accumulation_scheduler.going_to_accumulate_grad_batches()
+
+        if has_optimizer_step and going_to_accumulate_grad_batches and automatic_optimization:
+            raise MisconfigurationException(
+                'When overriding `LightningModule` optimizer_step with Trainer(..., automatic_optimization=True, ...), '
+                'accumulate_grad_batches should to be 1. It ensures this `optimizer_step` is called on every batch'
```
Comment (Borda) [inline pytorch_lightning/trainer/configuration_validator.py:88]:
```suggestion
                'When overriding `LightningModule` optimizer_step with `Trainer(..., automatic_optimization=True, ...)`,'
                ' `accumulate_grad_batches` should to be 1. It ensures this `optimizer_step` is called on every batch'
```

**Review by Borda — APPROVED** [review]
lgtm

File: pytorch_lightning/core/optimizer.py [hunk 11]
```diff
@@ -184,30 +227,19 @@ def dis_closure():
 
         if make_optimizer_step is None:
             make_optimizer_step = not self._should_accumulate
-
-        trainer = self._trainer
-        optimizer = self._optimizer
+        else:
+            if self._automatic_optimization:
+                raise MisconfigurationException(
```
Comment (Borda) [inline pytorch_lightning/core/optimizer.py:232]: add test for these cases...

File: pytorch_lightning/core/optimizer.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/core/optimizer.py:241]: Great idea !

File: pytorch_lightning/plugins/native_amp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/native_amp.py:75]: Hey both. 
For automatic_optimization, it is fine. It is called on training_step_and_backard.

For manual_optimization, we can't predict when people are going to make a `.step`. Therefore, the hook needs to be call after latest closure and `.step`.

However, I can move this to the precision plugin to reduce code.

File: pytorch_lightning/trainer/training_loop.py [hunk 12]
```diff
@@ -475,21 +475,34 @@ def _process_result(self, training_step_output, split_batch):
...
+        model_ref = self.trainer.get_model()
+
+        is_lbfgs = isinstance(optimizer, torch.optim.LBFGS)
+        using_native_amp = self.trainer.amp_backend == AMPType.NATIVE
+
+        # native amp + lbfgs is a no go right now
+        if using_native_amp and is_lbfgs:
+            raise MisconfigurationException(
+                'native PyTorch amp and lbfgs are not compatible.'
+                ' To request, please file a Github issue in PyTorch and tag @mcarilli')
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:487]: Yes. This is just legacy code from accelerators `optimizer_step`.

File: pytorch_lightning/plugins/native_amp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/native_amp.py:75]: This will be taken care in a next PR @awaelchli. I will need to clean out training_loop for `automatic_optimization` first, so it uses the same API as manual one.

File: pytorch_lightning/core/optimizer.py [hunk 13]
```diff
@@ -79,6 +102,31 @@ def _should_accumulate(self):
...
+                optimizer.step(closure=closure, *args, **kwargs)
+
+        trainer.train_loop.on_before_zero_grad(self)
+
+        model.optimizer_zero_grad(
+            trainer.current_epoch,
+            trainer.batch_idx,
+            optimizer,
+            self._optimizer_idx
+        )
```
Comment (rohitgr7) [inline pytorch_lightning/core/optimizer.py:128]: it won't be called if `enable_pl_optimizer=False`??

File: pytorch_lightning/core/optimizer.py [hunk 14]
```diff
@@ -60,17 +60,40 @@ def __init__(self,
...
 
     def _accumulated_batches_reached(self):
-        if self._use_accumulate_grad_batches_from_trainer:
-            accumulate_grad_batches = self._trainer.accumulate_grad_batches
-        else:
-            accumulate_grad_batches = self._accumulate_grad_batches
-        return (self._trainer.batch_idx + 1) % accumulate_grad_batches == 0
+        if self.accumulate_grad_batches is None:
+            return self._trainer.train_loop._accumulated_batches_reached()
+        return (self._trainer.batch_idx + 1) % self.accumulate_grad_batches == 0
```
Comment (rohitgr7) [inline pytorch_lightning/core/optimizer.py:96]: isn't `_accumulate_grad_batches` always None here? I can't it being set to a value anywhere.

File: pytorch_lightning/core/optimizer.py [hunk 15]
```diff
@@ -60,17 +60,40 @@ def __init__(self,
         self._trainer = None
         self._optimizer = optimizer
         self._accumulate_grad_batches = accumulate_grad_batches
```
Comment (rohitgr7) [inline pytorch_lightning/core/optimizer.py:62]:
```suggestion
        self.accumulate_grad_batches = accumulate_grad_batches
```

**Review by rohitgr7 — COMMENTED** [review]
Also, can you change this
https://github.com/PyTorchLightning/pytorch-lightning/blob/b00991efd8d6b7d1941d0eb3c1a499f95b4a3eea/pytorch_lightning/accelerators/accelerator.py#L163
to just
`if self.trainer.testing`

I'll close my https://github.com/PyTorchLightning/pytorch-lightning/pull/4982 then since these were the only 2 issues it's resolving.

File: pytorch_lightning/trainer/configuration_validator.py [hunk 16]
```diff
@@ -69,6 +69,25 @@ def __verify_train_loop_configuration(self, model):
...
             )
 
+        has_optimizer_step = is_overridden('optimizer_step', model)
+        enable_pl_optimizer = self.trainer._enable_pl_optimizer
+        automatic_optimization = self.trainer.train_loop.automatic_optimization
+        if has_optimizer_step and not enable_pl_optimizer and automatic_optimization:
+            log.warn(
+                "When overriding `LightningModule` optimizer_step with"
+                " `Trainer(..., enable_pl_optimizer=False, automatic_optimization=True, ...)`,"
+                " we won t be calling `.zero_grad` we can't assume when you call your `optimizer.step()`."
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/configuration_validator.py:79]:
```suggestion
                " we won't be calling `.zero_grad` we can't assume when you call your `optimizer.step()`."
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 17]
```diff
@@ -69,6 +69,25 @@ def __verify_train_loop_configuration(self, model):
...
+                " For Lightning to take care of it, please use `Trainer(enable_pl_optimizer=True)`."
+            )
+
+        going_to_accumulate_grad_batches = self.trainer.accumulation_scheduler.going_to_accumulate_grad_batches()
+
+        if has_optimizer_step and going_to_accumulate_grad_batches and automatic_optimization:
+            raise MisconfigurationException(
+                'When overriding `LightningModule` optimizer_step with `Trainer(automatic_optimization=True, ...)`,'
+                ' `accumulate_grad_batches` should to be 1. It ensures `optimizer_step` is called on every batch'
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/configuration_validator.py:88]: Yes, it will. 
However, when people are overriding optimizer_step. They might not pay attention to batch_idx indices.
They might do something like this. 
```
Trainer(accumulate_grad_batches=2)
...
def optimizer_step(...):
     batch_idx % 3 ==0: optimizer.step()
```
which will result in accumulated gradient of 6 and not 2 as expected.
It is why I found safer to make sure the `optimizer_step` is called on every batch_idx.

File: pytorch_lightning/core/optimizer.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/core/optimizer.py:128]: When people are using `automatic_optimization=True` and `enable_pl_optimizer=False` without overriding `optimizer_step`.

We will wrap their optimizer only the time to run step. So this function zero_grad will be called.
However, if they override `optimizer_step` with  Trainer(automatic_optimization=True, enable_pl_optimizer=False), they will get a warning we can't take care of zero_grad for them.

```
        if not isinstance(optimizer, LightningOptimizer):
            # wraps into LightingOptimizer only for running step
            optimizer = LightningOptimizer.to_lightning_optimizer(optimizer, self.trainer)
        optimizer.step(closure=optimizer_closure, *args, **kwargs)
```

File: pytorch_lightning/core/optimizer.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/core/optimizer.py:96]: Yes, _accumulate_grad_batches is None.
People will be able to set an accumulated value if they want.
```
optimizer = LightningOptimizer(optimizer, 3)

or 

optimizer.accumulate_grad_batches = 3
```

Or could even have random `.step`.
```
optimizer.step(make_optimzer_step=np.random.randint(100) > 90)
```

File: pytorch_lightning/core/optimizer.py (hunk 15 above)
Comment (justusschock) [reply pytorch_lightning/core/optimizer.py:62]: Actually I think the initial value should be set to the private attribute or at least the attribute should be initialised first

**Review by justusschock — APPROVED** [review]

**Review by williamFalcon — APPROVED** [review]
